pyvista/core/cell.py

```diff
@@ -176,8 +176,16 @@ def cast_to_unstructured_grid(self) -> pyvista.UnstructuredGrid:
           N Arrays:     0
 
         """
+        if self.type == CellType.POLYHEDRON:
+            # construct from faces
+            cell_ids = [self.n_faces]
+            for face in self.faces:
+                cell_ids.extend([len(face.point_ids)] + face.point_ids)
+            cell_ids.insert(0, len(cell_ids))
+        else:
+            cell_ids = [len(self.point_ids)] + list(range(len(self.point_ids)))
```

Comment:
I suspect this should also be:
```suggestion
            cell_ids = [len(self.point_ids)] + self.point_ids
```

Reply:
I _think_ for all other cell types, there should be no difference, but I'm not 100% sure.  @beroda do you have an opinion on this implementation?

Reply:
This suggestion does not work, as the connectivity is not correct since the `point_ids` are coming from the original mesh the cell was extracted from. Therefore, it can refers to `point_ids` that do not exist when declaring the `ugrid`.

With your suggestion, this simple example segfaults:
```python
from pyvista.examples import load_hexbeam

cell = load_hexbeam().get_cell(0)
cell.plot()
```

When looking closer at the connectivity array, it becomes clear:

**suggestion:**
```python
ugrid = cell.cast_to_unstructured_grid()
ugrid.cells
```
`array([ 8,  0,  2,  8,  7, 27, 36, 90, 81])`
```python
ugrid.n_points
```
`8`

**current implementation:**
```python
ugrid = cell.cast_to_unstructured_grid()
ugrid.cells
```
`array([8, 0, 1, 2, 3, 4, 5, 6, 7])`
```python
ugrid.n_points
```
`8`

Let me know if I was not clear.

Reply:
Hmmm this will need some testing with Polyhedron.  Thanks for the insight.  I had to change it to get it to work for that since it has a nested structure.

Reply:
Moreover, I am not sure whether the current implementation is correct for `polyhedron`.
This example does not show the cell correctly:
```python
import pyvista as pv
import numpy as np
from pyvista import CellType


ptIds = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
nodes = np.asarray(
    [
        [0.01, 0.01, 0.01],
        [0, 0.01, 0.01],
        [0, 0.0, 0.01],
        [0.01, 0.0, 0.01],
        [0.01, 0.0, 0.0],
        [0, 0.0, 0.0],
        [0, 0.01, 0.0],
        [0.01, 0.01, 0.0],
        [0.01, 0.01, 0.02],
        [0.00, 0.01, 0.02],
    ]
)

# polyhedron format [NElements, NFaces, Face1NPoints, Face1Point1, Face1Point2..., Face1PointN, FaceNNPoints,...]]
# in polygons and polyhedrons, face order/points order matters
polyhedron = np.asarray(
    [24, 5, 4, 2, 3, 4, 5, 4, 5, 6, 7, 4, 3, 4, 7, 3, 3, 5, 6, 2, 4, 6, 7, 3, 2]
)
polygon = np.asarray([4, 1, 9, 8, 0])

cells = np.hstack((polygon, polyhedron))
cell_type = [CellType.POLYGON, CellType.POLYHEDRON]
grid = pv.UnstructuredGrid(cells, cell_type, nodes)
grid.get_cell(1).plot()
```

I guess the implementation might be trickier than expected ...

Reply:
This should work IMO (sorry for the screenshot, it is late here :) ).
![image](https://github.com/pyvista/pyvista/assets/117822458/93faad55-1147-4b58-ae6e-18f57ba8f698)

Reply:
I think you are right.  I will test it.  These Polyhedrons are quite tricky.

Reply:
This also shows we really need an `__eq__` method for cells, mostly for testing.

Reply:
I am not sure whether an `__eq__`would solve this issue.

IMO, the issues we are facing with are related to the fact that you can construct an `UnstructuredGrid` with an erroneous connectivity (ie. the `point_ids` in the connectivity array can be higher than the number of points).

Correct me if I am wrong, but I think that no connectivity checks are performed in the `pyvista.UnstructuredGrid` constructor, nor for any `Dataset` with cells support. I think we should implement connectivity checks to catch such errors as early as possible (currently, it crashes only when plotting, which can confuse users on the failure reasons).

Reply:
We have several tests that do something like

```py
cell = grid.get_cell(0)
new_cell = cell.some_method()
assert cell == new_cell
```

This includes `cast_to_unstructured_grid`. This means that the current equality is not useful.  I tried to implement this for this PR and the `copy` method tests also started failing.

Failing early would also be good.

Reply:
But, a cell outside its mesh may no longer be equal to the original cell in the fullest sense in PyVista anyway.  Right now, `Cell` inherits from `DataObject`, which has a suspiciously `DataSet` focused `__eq__` method.

Reply:
> I think we should implement connectivity checks to catch such errors as early as possible (currently, it crashes only when plotting, which can confuse users on the failure reasons). If I am correct on this, I think I can find some time to propose a PR.

Not exactly the same context, but similar (structured grid consistency) discussion was [here](https://github.com/pyvista/pyvista/issues/4135#issuecomment-1470354274) with an explicit preference not to be stricter than VTK. Just FYI for "prior art".

Reply:
> This includes `cast_to_unstructured_grid`. This means that the current equality is not useful. I tried to implement this for this PR and the `copy` method tests also started failing.
> 
> Failing early would also be good.
I agree, this might need some investigation in another PR.

> Right now, `Cell` inherits from `DataObject`, which has a suspiciously `DataSet` focused `__eq__` method.
I think it confims the fact that `Cell` should have its own `__eq__`.

Reply:
> Not exactly the same context, but similar (structured grid consistency) discussion was [here](https://github.com/pyvista/pyvista/issues/4135#issuecomment-1470354274) with an explicit preference not to be stricter than VTK. Just FYI for "prior art".
It makes sense. My point was that for a user who is not a VTK expert and encounters such segfaults, it can be very tricky to understand where the problem comes from.
But I guess that a if a user tries to create the connectivity by himself, he followed some examples and is more susceptible to be aware of such issues as he gained VTK knowledge.